File: internal/domain/fix.go
// Package domain — FixPlan computes the minimal version bumps required
// to resolve known vulnerabilities across a snapshot. Pure function over
// (Snapshot + Advisories). No I/O.
//
// Algorithm:
//   - Group all dep advisories by (ecosystem, name).
//   - For each group, pick the highest FixedIn version (lexicographic
//     ordering — good enough for the npm/cargo/go semver-like strings
//     that dominate; PEP 440 / Maven coordinates are best-effort).
//   - The resulting target is the smallest single upgrade that clears
//     every advisory whose FixedIn is set.
//   - Advisories without FixedIn (rare — OSV usually has it) remain
//     "unresolved" and are surfaced separately so the user can decide
//     to vendor-patch or remove the dep.
package domain

import (
	"fmt"
	"strings"
)

// FixPlan is the result of BuildFixPlan: one entry per affected dep.
type FixPlan struct {
	Items []FixItem
}

// Empty reports whether the plan has nothing to do.
func (p FixPlan) Empty() bool {
	return len(p.Items) == 0
}

// FixItem is one dep that has at least one known vulnerability and a
// concrete upgrade target. Resolved is the subset of advisories whose
// FixedIn the upgrade clears; Unresolved are advisories with no
// upstream fix metadata.
type FixItem struct {
	Dep                  Dependency
	TargetVersion        string // "" when no FixedIn data on any advisory
	ResolvedAdvisories   []Advisory
	UnresolvedAdvisories []Advisory
}

// BuildFixPlan computes a FixPlan from a snapshot. Deps with no
// advisories are skipped. VEX-suppressed advisories are ignored so the
// fix planner doesn't propose upgrades for cleared CVEs.
func BuildFixPlan(snap Snapshot) FixPlan {
	var out FixPlan
	for _, d := range snap.Deps {
		if len(d.Advisories) == 0 {
			continue
		}
		var resolved, unresolved []Advisory
		var target string
		for _, a := range d.Advisories {
			if a.VEXSuppressed || a.FunctionUnreachable {
				continue
			}
			if a.FixedIn == "" {
				unresolved = append(unresolved, a)
				continue
			}
			resolved = append(resolved, a)
			if target == "" || compareFixVersion(a.FixedIn, target) > 0 {
				target = a.FixedIn
			}
		}
		if len(resolved) == 0 && len(unresolved) == 0 {
			continue
		}
		out.Items = append(out.Items, FixItem{
			Dep:                  d,
			TargetVersion:        target,
			ResolvedAdvisories:   resolved,
			UnresolvedAdvisories: unresolved,
		})
	}
	return out
}

// compareFixVersion compares two version strings. Returns -1 / 0 / 1.
// MVP: lexicographic with numeric-aware segment compare. Handles common
// semver shapes ("1.2.3", "1.2.10", "v1.2.3") correctly without pulling
// in golang.org/x/mod/semver. Falls back to byte compare for shapes
// outside semver (PEP 440 epoch markers, Maven qualifier strings).
func compareFixVersion(a, b string) int {
	a = strings.TrimPrefix(a, "v")
	b = strings.TrimPrefix(b, "v")
	as := strings.Split(a, ".")
	bs := strings.Split(b, ".")
	for i := 0; i < len(as) && i < len(bs); i++ {
		ax, aIsNum := parseLeadingInt(as[i])
		bx, bIsNum := parseLeadingInt(bs[i])
		if aIsNum && bIsNum {
			if ax != bx {
				if ax < bx {
					return -1
				}
				return 1
			}
			continue
		}
		if as[i] != bs[i] {
			if as[i] < bs[i] {
				return -1
			}
			return 1
		}
	}
	switch {
	case len(as) < len(bs):
		return -1
	case len(as) > len(bs):
		return 1
	}
	return 0
}

// isSafePkgName reports whether name only contains characters that
// are safe to interpolate into a shell command line without quoting:
//
//	[A-Za-z0-9._@/+-]
//
// Covers every legitimate package name across npm (incl. scoped
// `@scope/name`), PyPI, Cargo, Go modules (`golang.org/x/...`), Maven
// coordinates (`group:artifact` — `:` excluded; Maven path rendered
// separately), RubyGems, NuGet, Packagist, Gleam. Anything outside
// this charset (whitespace, `;`, `|`, `$`, backtick, quote, etc.)
// makes the name shell-unsafe and UpgradeCommand refuses to render.
func isSafePkgName(name string) bool {
	if name == "" || len(name) > 256 {
		return false
	}
	for i := 0; i < len(name); i++ {
		if !isSafeNameByte(name[i]) {
			return false
		}
	}
	return true
}

// isSafeNameByte is the per-byte ASCII allowlist used by isSafePkgName
// and isSafeVersion. ":" is allowed because Maven coords contain it,
// but the Maven branch of UpgradeCommand splits before interpolating.
func isSafeNameByte(c byte) bool {
	switch {
	case c >= 'a' && c <= 'z',
		c >= 'A' && c <= 'Z',
		c >= '0' && c <= '9':
		return true
	}
	switch c {
	case '.', '_', '-', '+', '/', '@', ':', '~':
		return true
	}
	return false
}

// isSafeVersion validates the targetVersion injected into upgrade
// commands. Stricter than name validation because we have less reason
// to allow exotic chars in versions; `^` and other range operators are
// excluded so an attacker who taints a FixedIn field can't smuggle
// `1.0.0; rm -rf /` past us. Accepts: digits, letters, `.+-_`. Covers
// every semver and PEP 440 shape we've seen in real OSV data.
func isSafeVersion(v string) bool {
	if v == "" || len(v) > 128 {
		return false
	}
	for i := 0; i < len(v); i++ {
		c := v[i]
		switch {
		case c >= 'a' && c <= 'z',
			c >= 'A' && c <= 'Z',
			c >= '0' && c <= '9':
			continue
		}
		switch c {
		case '.', '+', '-', '_':
			continue
		}
		return false
	}
	return true
}

// parseLeadingInt reads the leading numeric prefix of s. Returns (n,
// true) when s starts with at least one digit; (0, false) otherwise.
// Used by compareFixVersion to handle "1.2.10" > "1.2.9" correctly.
func parseLeadingInt(s string) (int, bool) {
	n := 0
	any := false
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c < '0' || c > '9' {
			break
		}
		n = n*10 + int(c-'0')
		any = true
	}
	return n, any
}

// UpgradeCommand returns the ecosystem-appropriate upgrade shell command
// for dep. When targetVersion is non-empty, the command pins to that
// version; otherwise it falls back to the ecosystem's "latest" verb.
// Returns "" when no command shape is known.
//
// Security: package names and target versions are validated against a
// strict identifier charset before being interpolated. The output is
// intended to be safe to `eval` or pipe to `sh` (e.g. via `aegis fix
// --script | sh`). Any name/version containing shell metacharacters,
// whitespace, or anything outside the per-ecosystem identifier alphabet
// causes UpgradeCommand to return "" — the caller surfaces the offending
// dep so the user can investigate manually rather than executing a
// crafted upgrade line.
func UpgradeCommand(dep Dependency, targetVersion string) string {
	if !isSafePkgName(dep.Name) {
		return ""
	}
	if targetVersion != "" && !isSafeVersion(targetVersion) {
		return ""
	}
	name := dep.Name
	pinned := targetVersion != ""
	switch dep.Ecosystem {
	case EcoNpm:
		if pinned {
			return fmt.Sprintf("npm install %s@%s", name, targetVersion)
		}
		return fmt.Sprintf("npm install %s@latest", name)
	case EcoPyPI:
		if pinned {
			return fmt.Sprintf("pip install %s==%s", name, targetVersion)
		}
		return fmt.Sprintf("pip install --upgrade %s", name)
	case EcoRubyGems:
		if pinned {
			return fmt.Sprintf("bundle update --conservative %s  # target: %s", name, targetVersion)
		}
		return fmt.Sprintf("bundle update %s", name)
	case EcoCrates:
		cratesName := strings.ReplaceAll(name, "/", "-")
		if pinned {
			return fmt.Sprintf("cargo update -p %s --precise %s", cratesName, targetVersion)
		}
		return fmt.Sprintf("cargo update -p %s", cratesName)
	case EcoGo:
		if pinned {
			return fmt.Sprintf("go get %s@%s", name, targetVersion)
		}
		return fmt.Sprintf("go get %s@latest", name)
	case EcoMaven:
		parts := strings.SplitN(name, ":", 2)
		if len(parts) == 2 {
			if pinned {
				return fmt.Sprintf("mvn versions:set-property -Dproperty=%s.version -DnewVersion=%s", parts[1], targetVersion)
			}
			return fmt.Sprintf("mvn versions:use-latest-versions -Dincludes=%s:%s", parts[0], parts[1])
		}
	case EcoPackagist:
		if pinned {
			return fmt.Sprintf("composer require %s:%s", name, targetVersion)
		}
		return fmt.Sprintf("composer update %s", name)
	case EcoNuGet:
		if pinned {
			return fmt.Sprintf("dotnet add package %s --version %s", name, targetVersion)
		}
		return fmt.Sprintf("dotnet add package %s", name)
	case EcoGleam:
		return "gleam deps update"
	}
	return ""
}

File: crates/aegis-domain/src/fix.rs
//! Fix planning — the minimal version bumps that resolve known
//! vulnerabilities. Pure port of `domain/fix.go`. No I/O.
//!
//! Group advisories per dep; the target version is the highest `fixed_in`
//! that is a forward upgrade from the installed version (a `fixed_in` at or
//! below the install is a backport on an older range, not a valid fix — never
//! a downgrade). Advisories without `fixed_in`, or whose `fixed_in` isn't a
//! forward upgrade, stay unresolved for the user to handle manually.

use std::cmp::Ordering;

use crate::{Advisory, Dependency, Ecosystem};

/// One dep with at least one advisory and a computed upgrade target.
#[derive(Debug, Clone)]
pub struct FixItem {
    pub dep: Dependency,
    /// Highest forward `fixed_in`; empty when no advisory has usable fix data.
    pub target_version: String,
    /// Advisories the upgrade to `target_version` clears.
    pub resolved: Vec<Advisory>,
    /// Advisories with no forward fix (no `fixed_in`, or a backport/downgrade).
    pub unresolved: Vec<Advisory>,
}

/// Compute the fix plan from (dependency, its advisories) pairs. Deps with no
/// advisories are skipped.
pub fn build_fix_plan(deps: &[(Dependency, Vec<Advisory>)]) -> Vec<FixItem> {
    let mut out = Vec::new();
    for (dep, advisories) in deps {
        if advisories.is_empty() {
            continue;
        }
        let mut resolved = Vec::new();
        let mut unresolved = Vec::new();
        let mut target = String::new();
        for a in advisories {
            if a.fixed_in.is_empty() {
                unresolved.push(a.clone());
                continue;
            }
            // Only a forward upgrade clears the advisory; never a downgrade.
            if compare_fix_version(&a.fixed_in, &dep.version) != Ordering::Greater {
                unresolved.push(a.clone());
                continue;
            }
            if target.is_empty() || compare_fix_version(&a.fixed_in, &target) == Ordering::Greater {
                target = a.fixed_in.clone();
            }
            resolved.push(a.clone());
        }
        if resolved.is_empty() && unresolved.is_empty() {
            continue;
        }
        out.push(FixItem {
            dep: dep.clone(),
            target_version: target,
            resolved,
            unresolved,
        });
    }
    out
}

/// Compare two version strings. When both parse as semver, uses full semver
/// precedence — including pre-release ordering, so `1.2.3-rc.1 < 1.2.3` (the
/// naive per-segment compare got this backwards). Falls back to a
/// numerically-aware per-dotted-segment compare for non-semver shapes
/// (date tags, `latest`, calendar versions). Handles a leading `v`.
pub fn compare_fix_version(a: &str, b: &str) -> Ordering {
    // Prefer real semver precedence when both sides parse.
    if let (Some(va), Some(vb)) = (
        crate::semver::Version::parse(a),
        crate::semver::Version::parse(b),
    ) {
        return va.cmp(&vb);
    }
    let a = a.strip_prefix('v').unwrap_or(a);
    let b = b.strip_prefix('v').unwrap_or(b);
    let (as_, bs): (Vec<&str>, Vec<&str>) = (a.split('.').collect(), b.split('.').collect());
    for (x, y) in as_.iter().zip(bs.iter()) {
        match (parse_leading_int(x), parse_leading_int(y)) {
            (Some(ax), Some(by)) => {
                if ax != by {
                    return ax.cmp(&by);
                }
            }
            _ => {
                if x != y {
                    return x.cmp(y);
                }
            }
        }
    }
    as_.len().cmp(&bs.len())
}

/// Leading numeric prefix of `s`, or `None` if it doesn't start with a digit.
fn parse_leading_int(s: &str) -> Option<u64> {
    let digits: String = s.chars().take_while(|c| c.is_ascii_digit()).collect();
    digits.parse().ok()
}

/// True when `name` contains only chars safe to interpolate into a shell
/// upgrade command without quoting. Mirrors `isSafePkgName`.
fn is_safe_pkg_name(name: &str) -> bool {
    if name.is_empty() || name.len() > 256 {
        return false;
    }
    name.bytes().all(|c| {
        c.is_ascii_alphanumeric()
            || matches!(c, b'.' | b'_' | b'-' | b'+' | b'/' | b'@' | b':' | b'~')
    })
}

/// Stricter than [`is_safe_pkg_name`] — range operators (`^`, `~`, …) excluded
/// so a tainted `fixed_in` can't smuggle shell metacharacters. Mirrors
/// `isSafeVersion`.
fn is_safe_version(v: &str) -> bool {
    if v.is_empty() || v.len() > 128 {
        return false;
    }
    v.bytes()
        .all(|c| c.is_ascii_alphanumeric() || matches!(c, b'.' | b'+' | b'-' | b'_'))
}

/// The ecosystem-appropriate upgrade command for `dep`, pinned to
/// `target_version` when given. `None` when no command shape is known or the
/// name/version fails the shell-safety check. Output is safe to pipe to `sh`.
/// Mirrors `UpgradeCommand`.
pub fn upgrade_command(dep: &Dependency, target_version: &str) -> Option<String> {
    if !is_safe_pkg_name(&dep.name) {
        return None;
    }
    if !target_version.is_empty() && !is_safe_version(target_version) {
        return None;
    }
    let name = &dep.name;
    let pinned = !target_version.is_empty();
    let t = target_version;
    let cmd = match dep.ecosystem {
        Ecosystem::Npm => {
            if pinned {
                format!("npm install {name}@{t}")
            } else {
                format!("npm install {name}@latest")
            }
        }
        Ecosystem::PyPI => {
            if pinned {
                format!("pip install {name}=={t}")
            } else {
                format!("pip install --upgrade {name}")
            }
        }
        Ecosystem::RubyGems => {
            if pinned {
                format!("bundle update --conservative {name}  # target: {t}")
            } else {
                format!("bundle update {name}")
            }
        }
        Ecosystem::Crates => {
            let cn = name.replace('/', "-");
            if pinned {
                format!("cargo update -p {cn} --precise {t}")
            } else {
                format!("cargo update -p {cn}")
            }
        }
        Ecosystem::Go => {
            if pinned {
                format!("go get {name}@{t}")
            } else {
                format!("go get {name}@latest")
            }
        }
        Ecosystem::Maven => {
            let (group, artifact) = name.split_once(':')?;
            if pinned {
                format!("mvn versions:set-property -Dproperty={artifact}.version -DnewVersion={t}")
            } else {
                format!("mvn versions:use-latest-versions -Dincludes={group}:{artifact}")
            }
        }
        Ecosystem::Packagist => {
            if pinned {
                format!("composer require {name}:{t}")
            } else {
                format!("composer update {name}")
            }
        }
        Ecosystem::NuGet => {
            if pinned {
                format!("dotnet add package {name} --version {t}")
            } else {
                format!("dotnet add package {name}")
            }
        }
        Ecosystem::Pub => {
            if pinned {
                format!("dart pub upgrade {name}  # target: {t} (pin in pubspec.yaml)")
            } else {
                format!("dart pub upgrade {name}")
            }
        }
        Ecosystem::CocoaPods => {
            if pinned {
                format!("pod update {name}  # target: {t} (pin in Podfile)")
            } else {
                format!("pod update {name}")
            }
        }
        Ecosystem::Cpan => {
            if pinned {
                format!("cpanm {name}@{t}")
            } else {
                format!("cpanm {name}")
            }
        }
        Ecosystem::Hackage => {
            if pinned {
                format!("cabal install --constraint=\"{name} =={t}\"")
            } else {
                format!("cabal install {name}")
            }
        }
        Ecosystem::Cran => {
            if pinned {
                format!("R -e 'remotes::install_version(\"{name}\", \"{t}\")'")
            } else {
                format!("R -e 'install.packages(\"{name}\")'")
            }
        }
        Ecosystem::Hex => {
            // hex.pm serves Gleam + Elixir; the dep doesn't record which, so
            // emit a manual hint naming both toolchains rather than guessing.
            if pinned {
                format!("# upgrade {name} to {t}: `mix deps.update {name}` (Elixir) or `gleam update {name}` (Gleam)")
            } else {
                format!("# upgrade {name}: `mix deps.update {name}` (Elixir) or `gleam update {name}` (Gleam)")
            }
        }
        // No known upgrade shape.
        Ecosystem::SwiftPM
        | Ecosystem::Neovim
        | Ecosystem::Aur
        | Ecosystem::Conan
        | Ecosystem::Nix
        | Ecosystem::Julia
        | Ecosystem::Conda
        | Ecosystem::Nimble => return None,
    };
    Some(cmd)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Severity;

    fn adv(id: &str, fixed_in: &str) -> Advisory {
        Advisory {
            id: id.into(),
            severity: Severity::High,
            fixed_in: fixed_in.into(),
            ..Default::default()
        }
    }
    fn dep(eco: Ecosystem, name: &str, version: &str) -> Dependency {
        Dependency {
            ecosystem: eco,
            name: name.into(),
            version: version.into(),
            ..Default::default()
        }
    }

    #[test]
    fn compare_numeric_aware() {
        assert_eq!(compare_fix_version("1.2.10", "1.2.9"), Ordering::Greater);
        assert_eq!(compare_fix_version("v1.0.0", "1.0.0"), Ordering::Equal);
        assert_eq!(compare_fix_version("1.0", "1.0.1"), Ordering::Less);
        assert_eq!(compare_fix_version("2.0.0", "1.9.9"), Ordering::Greater);
    }

    #[test]
    fn compare_orders_prerelease_below_release() {
        // The naive per-segment compare got this backwards (treated the extra
        // `-rc.1` segment as "greater"); semver precedence puts it lower.
        assert_eq!(compare_fix_version("1.2.3-rc.1", "1.2.3"), Ordering::Less);
        assert_eq!(
            compare_fix_version("1.2.3", "1.2.3-rc.1"),
            Ordering::Greater
        );
        assert_eq!(
            compare_fix_version("1.0.0-alpha", "1.0.0-beta"),
            Ordering::Less
        );
        // Non-semver shapes still fall back to the byte/segment compare.
        assert_eq!(compare_fix_version("latest", "latest"), Ordering::Equal);
    }

    #[test]
    fn plan_does_not_treat_prerelease_as_forward_fix() {
        // Installed 1.2.3; a `fixed_in` of a *prerelease* of the same release
        // is not a forward upgrade → stays unresolved (was wrongly "resolved"
        // before the semver fix).
        let d = dep(Ecosystem::Npm, "pkg", "1.2.3");
        let plan = build_fix_plan(&[(d, vec![adv("GHSA-x", "1.2.3-rc.1")])]);
        assert_eq!(plan.len(), 1);
        assert!(plan[0].target_version.is_empty());
        assert_eq!(plan[0].unresolved.len(), 1);
        assert!(plan[0].resolved.is_empty());
    }

    #[test]
    fn plan_picks_highest_forward_fix() {
        let d = dep(Ecosystem::Npm, "lodash", "4.17.4");
        let advs = vec![adv("GHSA-a", "4.17.11"), adv("GHSA-b", "4.17.21")];
        let plan = build_fix_plan(&[(d, advs)]);
        assert_eq!(plan.len(), 1);
        assert_eq!(plan[0].target_version, "4.17.21");
        assert_eq!(plan[0].resolved.len(), 2);
        assert!(plan[0].unresolved.is_empty());
    }

    #[test]
    fn plan_never_downgrades() {
        // fixed_in <= installed → unresolved, no target.
        let d = dep(Ecosystem::Npm, "x", "5.0.0");
        let advs = vec![adv("GHSA-old", "4.9.0")];
        let plan = build_fix_plan(&[(d, advs)]);
        assert_eq!(plan.len(), 1);
        assert_eq!(plan[0].target_version, "");
        assert_eq!(plan[0].unresolved.len(), 1);
    }

    #[test]
    fn plan_advisory_without_fixed_in_unresolved() {
        let d = dep(Ecosystem::Npm, "x", "1.0.0");
        let plan = build_fix_plan(&[(d, vec![adv("GHSA-c", "")])]);
        assert_eq!(plan[0].unresolved.len(), 1);
        assert_eq!(plan[0].target_version, "");
    }

    #[test]
    fn upgrade_commands_per_ecosystem() {
        assert_eq!(
            upgrade_command(&dep(Ecosystem::Npm, "lodash", "0"), "4.17.21").as_deref(),
            Some("npm install lodash@4.17.21")
        );
        assert_eq!(
            upgrade_command(&dep(Ecosystem::PyPI, "requests", "0"), "2.31.0").as_deref(),
            Some("pip install requests==2.31.0")
        );
        assert_eq!(
            upgrade_command(&dep(Ecosystem::Crates, "serde", "0"), "1.0.0").as_deref(),
            Some("cargo update -p serde --precise 1.0.0")
        );
        assert_eq!(
            upgrade_command(&dep(Ecosystem::Go, "github.com/x/y", "0"), "v1.2.0").as_deref(),
            Some("go get github.com/x/y@v1.2.0")
        );
        // unpinned falls back to latest verbs
        assert_eq!(
            upgrade_command(&dep(Ecosystem::Npm, "x", "0"), "").as_deref(),
            Some("npm install x@latest")
        );
    }

    #[test]
    fn maven_coord_split() {
        assert_eq!(
            upgrade_command(
                &dep(Ecosystem::Maven, "com.google.guava:guava", "0"),
                "33.0"
            )
            .as_deref(),
            Some("mvn versions:set-property -Dproperty=guava.version -DnewVersion=33.0")
        );
    }

    #[test]
    fn refuses_shell_unsafe_name_or_version() {
        assert!(upgrade_command(&dep(Ecosystem::Npm, "x; rm -rf /", "0"), "1.0.0").is_none());
        assert!(upgrade_command(&dep(Ecosystem::Npm, "x", "0"), "1.0.0; evil").is_none());
    }

    #[test]
    fn no_command_shape_returns_none() {
        assert!(upgrade_command(&dep(Ecosystem::SwiftPM, "x", "0"), "1.0.0").is_none());
    }
}
